miner/proposal_builder.go

```diff
@@ -144,175 +195,85 @@ func WithNetworkDelay(delay time.Duration) Opt {
 	}
 }
 
-func WithMinGoodAtxPct(pct int) Opt {
-	return func(pb *ProposalBuilder) {
-		pb.cfg.goodAtxPct = pct
-	}
-}
-
-func withOracle(o proposalOracle) Opt {
+func WithMinGoodAtxPercent(percent int) Opt {
 	return func(pb *ProposalBuilder) {
-		pb.proposalOracle = o
+		pb.cfg.GoodAtxPercent = percent
 	}
 }
 
-func withNonceFetcher(nf nonceFetcher) Opt {
-	return func(pb *ProposalBuilder) {
-		pb.nonceFetcher = nf
-	}
-}
-
-// NewProposalBuilder creates a struct of block builder type.
-func NewProposalBuilder(
-	ctx context.Context,
+// New creates a struct of block builder type.
+func New(
 	clock layerClock,
 	signer *signing.EdSigner,
-	vrfSigner *signing.VRFSigner,
 	cdb *datastore.CachedDB,
 	publisher pubsub.Publisher,
 	trtl votesEncoder,
-	beaconProvider system.BeaconGetter,
 	syncer system.SyncStateProvider,
 	conState conservativeState,
 	opts ...Opt,
 ) *ProposalBuilder {
-	sctx, cancel := context.WithCancel(ctx)
+	ctx, cancel := context.WithCancel(context.Background())
 	pb := &ProposalBuilder{
-		logger:         log.NewNop(),
-		ctx:            sctx,
-		cancel:         cancel,
-		signer:         signer,
-		clock:          clock,
-		cdb:            cdb,
-		publisher:      publisher,
-		tortoise:       trtl,
-		beaconProvider: beaconProvider,
-		syncer:         syncer,
-		conState:       conState,
+		logger:    log.NewNop(),
+		ctx:       ctx,
```

Comment:
Why keep the context in a struct instead of accepting a context from the caller in `Start()` method? It would stop when the passed context is canceled (app shutdown).

The `Stop()` could become
```go
func (pb *ProposalBuilder) WaitToStop() {
	_ = pb.eg.Wait()
}
```

---

I prefer to make Start/Run methods blocking and offload putting it onto a goroutine on the caller but this requires more refactoring of the existing codebase. Something like:
```go
var eg errgroup.Group
eg.Go(func() error { return proposalBuilder.Start(ctx) }
// start other "services"
eg.Go(func() error { return other.Start(ctx) }
...
eg.Wait() 
```

- `eg.Wait()` blocks waiting for all "services" to complete, there is no need to call Stop(), Close() etc.
- no need to store `ctx`, `cancel` and `eg` in service structs. `eg` is a local variable

WDYT?

Reply:
i didn't change as it was old code and didn't seem very important but in general i like two approaches:
- Run(ctx) the way you described
- Start() / Stop() and ctx initialized locally , so that lifecycle is managed by this object without external dependencies

Start/Stop has an "advantage" that it can be aware of specific order for Stop. will refactor to use Run(ctx)

Reply:
> I agree the service should not have a context.Context as a field

the advantage of storing context in Start/Stop approach is that context can be passed down the stack. so if some method makes a network request or generally stoppable we can pass context as is. with a struct channel there will be a need for additional context

Reply:
From the official documentation (https://pkg.go.dev/context):

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:

For the previous design I would have probably done something like the following:

```
func (pb *proposalBuilder) Start() error {
    if pb.stop != nil {
        return errors.New("service already started")
    }

    ctx, cancel := context.WithCancel(context.Background)
    pb.stop = cancel
    pb.eg.Go(func() {
       return pb.longRunningTask1(ctx, ...)
    })
    pb.eg.Go(func() {
       return pb.longRunningTask2(ctx, ...)
    })
   return nil
}

func (pb *proposalBuilder) Close() error {
    pb.stop()
    return pb.eg.Wait()
}
```

This way only a reference to `context.CancelFunc` is held.